{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Import Libraries\n",
    "#\n",
    "\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "from plotly.subplots import make_subplots\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "utilsPath = r'S:\\U_Proteomica\\UNIDAD\\software\\MacrosRafa\\data\\Metabolomics\\PESA_Integromics\\Data\\utils'\n",
    "if utilsPath not in sys.path:\n",
    "    sys.path.append(utilsPath)\n",
    "\n",
    "from myLog import myLog\n",
    "from PlotEDA import PlotEDA\n",
    "from PlotMV import PlotMV\n",
    "from PCA_UMAP import PCA_UMAP, PCA_Var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Constants\n",
    "#\n",
    "\n",
    "MVF_thr = 0.2 # Filter by missing value (IS non detected)\n",
    "MV5_thr = 0.2 # Filter by -5 imputed missing value (feature not detected)\n",
    "MVO_thr = 0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Set constants\n",
    "#\n",
    "\n",
    "workingPath = r\"S:\\U_Proteomica\\UNIDAD\\software\\MacrosRafa\\data\\Metabolomics\\PESA_Integromics\\Data\\Metabolomics\\AWHS\"\n",
    "pos_info_path = os.path.join(workingPath, 'OriginalFiles', 'RBR_LOG_POS_INFO.xlsx')\n",
    "neg_info_path = os.path.join(workingPath, 'OriginalFiles', 'RBR_LOG_NEG_INFO.xlsx')\n",
    "\n",
    "fileSummaryConc = os.path.join(workingPath, 'WorkingFiles', 'Plots', 'SummaryPlotsConc.html')\n",
    "fileMSV = os.path.join(workingPath, 'WorkingFiles', 'Plots', 'SummaryMSV.html')\n",
    "fileSummaryMS1 = os.path.join(workingPath, 'WorkingFiles', 'Plots', 'SummaryPlotsMS1.html')\n",
    "fileSummaryMS2S = os.path.join(workingPath, 'WorkingFiles', 'Plots', 'SummaryPlotsMS2S.html')\n",
    "fileSummaryMS2R = os.path.join(workingPath, 'WorkingFiles', 'Plots', 'SummaryPlotsMS2R.html')\n",
    "filePCAMS1 = os.path.join(workingPath, 'WorkingFiles', 'Plots', 'PCAPlotsMS1.html')\n",
    "filePCAMS2 = os.path.join(workingPath, 'WorkingFiles', 'Plots', 'PCAPlotsMS2.html')\n",
    "if os.path.exists(fileSummaryConc): os.remove(fileSummaryConc)\n",
    "if os.path.exists(fileMSV): os.remove(fileMSV)\n",
    "if os.path.exists(fileSummaryMS1): os.remove(fileSummaryMS1)\n",
    "if os.path.exists(fileSummaryMS2S): os.remove(fileSummaryMS2S)\n",
    "if os.path.exists(fileSummaryMS2R): os.remove(fileSummaryMS2R)\n",
    "if os.path.exists(filePCAMS1): os.remove(filePCAMS1)\n",
    "if os.path.exists(filePCAMS2): os.remove(filePCAMS2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start Session\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Set logging\n",
    "#\n",
    "\n",
    "logw = myLog(os.path.join(workingPath,'WorkingFiles', 'info.log'))\n",
    "logw('Start Session')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Read sheets on dataframes\n",
    "#\n",
    "\n",
    "xmp = pd.read_excel(pos_info_path, sheet_name='fid2log', header=[0,1], index_col=0) # Positive-Concentration\n",
    "mp2i = pd.read_excel(pos_info_path, sheet_name='fid2LipidInfo')\n",
    "# posS2sn = pd.read_excel(pos_info_path, sheet_name='sample2Seqn')\n",
    "\n",
    "xmn = pd.read_excel(neg_info_path, sheet_name='fid2log', header=[0,1], index_col=0, na_values=' None') # Negative-Concentration\n",
    "mn2i = pd.read_excel(neg_info_path, sheet_name='fid2LipidInfo')\n",
    "# negS2sn = pd.read_excel(neg_info_path, sheet_name='sample2Seqn')\n",
    "\n",
    "mdata = pd.read_csv(r'S:\\U_Proteomica\\UNIDAD\\software\\MacrosRafa\\data\\Metabolomics\\PESA_Integromics\\Data\\Metadata\\AWHS\\WorkingFiles\\main_metadata.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Generate m2info.tsv joining positive and negative\n",
    "#\n",
    "\n",
    "m2i = pd.concat([mp2i, mn2i])\n",
    "m2i['mode'] = ['POS' if i[0]=='P' else 'NEG' for i in m2i['fid']]\n",
    "m2i.to_csv(os.path.join(workingPath, 'WorkingFiles', 'm2info.tsv'), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Generate Xm_MS1.tsv and Xm_MS2.tsv\n",
    "# WARNING!!!!! for concentration some values are inf (We only use them for representation)\n",
    "\n",
    "# CONC_MS1.tsv\n",
    "xmp1 = xmp.loc[:, pd.IndexSlice['P']].T.dropna(axis=1, how='all')\n",
    "cmp1 = 10**xmp1\n",
    "xmn1 = xmn.loc[:, pd.IndexSlice['P']].T.dropna(axis=1, how='all')\n",
    "cmn1 = 10**xmn1\n",
    "\n",
    "xm1 = pd.merge(\n",
    "    xmp1,\n",
    "    xmn1,\n",
    "    how='outer',\n",
    "    on='fid'\n",
    ")\n",
    "\n",
    "cm1 = pd.merge(\n",
    "    cmp1,\n",
    "    cmn1,\n",
    "    how='outer',\n",
    "    on='fid'\n",
    ")\n",
    "\n",
    "xm1.to_csv(os.path.join(workingPath, 'WorkingFiles', 'Xm_MS1.tsv'), sep='\\t', index=True)\n",
    "\n",
    "# CONC_MS2.tsv\n",
    "xmp2 = xmp.loc[:, pd.IndexSlice['FRAG']].T.dropna(axis=1, how='all')\n",
    "cmp2 = 10**xmp2\n",
    "xmn2 = xmn.loc[:, pd.IndexSlice['FAS']].T.dropna(axis=1, how='all')\n",
    "cmn2 = 10**xmn2\n",
    "\n",
    "xm2 = pd.merge(\n",
    "    xmp2,\n",
    "    xmn2,\n",
    "    how='outer',\n",
    "    on='fid'\n",
    ")\n",
    "\n",
    "cm2 = pd.merge(\n",
    "    cmp2,\n",
    "    cmn2,\n",
    "    how='outer',\n",
    "    on='fid'\n",
    ")\n",
    "\n",
    "xm2.to_csv(os.path.join(workingPath, 'WorkingFiles', 'Xm_MS2.tsv'), sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of features MS1: 849\n",
      "Number of features MS1 (Pos.): 653\n",
      "Number of features MS1 (Neg.): 196\n",
      "Number of features MS2: 805\n",
      "Number of features MS2 (Pos.): 650\n",
      "Number of features MS2 (Neg.): 155\n"
     ]
    }
   ],
   "source": [
    "logw(f\"Number of features MS1: {cm1.shape[1]}\")\n",
    "logw(f\"Number of features MS1 (Pos.): {cmp1.shape[1]}\")\n",
    "logw(f\"Number of features MS1 (Neg.): {cmn1.shape[1]}\")\n",
    "logw(f\"Number of features MS2: {cm2.shape[1]}\")\n",
    "logw(f\"Number of features MS2 (Pos.): {cmp2.shape[1]}\")\n",
    "logw(f\"Number of features MS2 (Neg.): {cmn2.shape[1]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract from metadata subtable with observations studied in metabolomics\n",
    "mdatam = mdata[np.isin(mdata['Seqn'], cm1.index)].copy()\n",
    "mdatam['AllElem'] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Plot concentration distribution\n",
    "#\n",
    "palette = ['#636EFA', '#EF553B', '#00CC96', '#AB63FA', '#FFA15A', '#19D3F3', '#FF6692', '#B6E880']\n",
    "\n",
    "\n",
    "column = 'Metabo_Batch'\n",
    "valMax = 300\n",
    "dfs = [cm1, cm2]\n",
    "namedfs = ['MS1 Pos/Neg', 'MS2 Pos/Neg']\n",
    "\n",
    "def saveToFile(fig, file):\n",
    "    with open(file, 'a') as f:\n",
    "        f.write(fig.to_html(full_html=False, include_plotlyjs='cdn', default_height='50%', default_width='80%'))\n",
    "\n",
    "\n",
    "def PlotConcentration(dfs, namedfs, column, valMax=500, file=False):\n",
    "\n",
    "    fig = make_subplots(rows=2, cols=2, shared_xaxes=True, vertical_spacing=0.02, horizontal_spacing=0.05, subplot_titles=namedfs)\n",
    "    for ndf,df in enumerate(dfs):\n",
    "        for n,value in enumerate(sorted(set(mdatam[column]))):\n",
    "            data = df.loc[mdatam['Seqn'][mdatam[column]==value], :].to_numpy().flatten() \n",
    "            data = data[data<valMax]\n",
    "            fig.add_trace(go.Histogram(\n",
    "                x=data,\n",
    "                name=value,\n",
    "                opacity=0.75,\n",
    "                marker_color=palette[n],\n",
    "                showlegend=True if ndf==0 else False,\n",
    "                histnorm='probability density',\n",
    "                xbins=dict(size=1),\n",
    "                #hoverinfo='skip'\n",
    "            ), row=1, col=1+ndf)\n",
    "\n",
    "            fig.add_trace(go.Box(\n",
    "                x=data,\n",
    "                name=value,\n",
    "                opacity=0.75,\n",
    "                marker_color=palette[n],\n",
    "                showlegend=False,\n",
    "                hoverinfo='skip'\n",
    "            ), row=2, col=1+ndf)\n",
    "\n",
    "    fig.update_layout(\n",
    "        title=f\"Concentration distribution by {column}\",\n",
    "        barmode='overlay'\n",
    "    )\n",
    "        \n",
    "    fig.show() if not file else saveToFile(fig, file)\n",
    "\n",
    "PlotConcentration([cm1, cm2], ['MS1 Pos/Neg', 'MS2 Pos/Neg'], 'AllElem', valMax=300, file=fileSummaryConc)\n",
    "PlotConcentration([cmp1, cmn1], ['MS1 Pos', 'MS1 Neg'], 'AllElem', valMax=300, file=fileSummaryConc)\n",
    "PlotConcentration([cmp2, cmn2], ['MS2 Pos', 'MS2 Neg'], 'AllElem', valMax=300, file=fileSummaryConc)\n",
    "PlotConcentration([cm1, cm2], ['MS1 Pos/Neg', 'MS2 Pos/Neg'], 'Metabo_Batch', valMax=300, file=fileSummaryConc)\n",
    "# PlotConcentration([cmp1, cmn1], ['MS1 Pos', 'MS1 Neg'], 'Cohort', valMax=300, file=fileSummary)\n",
    "# PlotConcentration([cmp2, cmn2], ['MS2 Pos', 'MS2 Neg'], 'Cohort', valMax=300, file=fileSummary)\n",
    "PlotConcentration([cm1, cm2], ['MS1 Pos/Neg', 'MS2 Pos/Neg'], 'Group', valMax=300, file=fileSummaryConc)\n",
    "# PlotConcentration([cmp1, cmn1], ['MS1 Pos', 'MS1 Neg'], 'Group', valMax=300, file=fileSummary)\n",
    "# PlotConcentration([cmp2, cmn2], ['MS2 Pos', 'MS2 Neg'], 'Group', valMax=300, file=fileSummary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Plot a scatter indicating for each feature % of missing value that were imputed with -5.\n",
    "#\n",
    "\n",
    "m2im1 = m2i[np.isin(m2i['fid'], xm1.columns)]\n",
    "m2im2 = m2i[np.isin(m2i['fid'], xm2.columns)]\n",
    "\n",
    "snc = mdatam['Seqn'][mdatam['Group']=='C']\n",
    "snd = mdatam['Seqn'][mdatam['Group']=='D']\n",
    "\n",
    "fig = make_subplots(rows=1, cols=2, subplot_titles=['MS1', 'MS2'])\n",
    "\n",
    "col=1\n",
    "for xm,m2im in zip([xm1, xm2], [m2im1, m2im2]):\n",
    "\n",
    "    # fig = go.Figure()\n",
    "    fig.add_trace(go.Scatter(\n",
    "        x=(xm.loc[snc, m2im['fid'][m2im['mode']=='POS']]==-5).sum()/len(snc),\n",
    "        y=(xm.loc[snd, m2im['fid'][m2im['mode']=='POS']]==-5).sum()/len(snd),\n",
    "        mode='markers',\n",
    "        name='POS',\n",
    "        marker_color='#636EFA',\n",
    "        showlegend=False if col==2 else True\n",
    "    ), row=1, col=col)\n",
    "    fig.add_trace(go.Scatter(\n",
    "        x=(xm.loc[snc, m2im['fid'][m2im['mode']=='NEG']]==-5).sum()/len(snc),\n",
    "        y=(xm.loc[snd, m2im['fid'][m2im['mode']=='NEG']]==-5).sum()/len(snd),\n",
    "        mode='markers',\n",
    "        name='NEG',\n",
    "        marker_color='#EF553B',\n",
    "        showlegend=False if col==2 else True\n",
    "    ), row=1, col=col)\n",
    "    fig.add_trace(go.Scatter(\n",
    "        x=[0,1], y=[0,1], line=dict(dash='dash', width=0.5, color='black'), opacity=0.7, showlegend=False, mode='lines'\n",
    "    ), row=1, col=col)\n",
    "    col+=1\n",
    "\n",
    "fig.update_xaxes(title='Control')\n",
    "fig.update_yaxes(title='Disease')\n",
    "fig.update_layout(title='Ratio of missing values imputed with -5 in Control vs Disease')\n",
    "saveToFile(fig, fileSummaryConc)#fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing Values\n",
      "\n",
      "MS1\n",
      "All values: 295452\n",
      "Values imputed with -5: 14543 (4.92%)\n",
      "Non-imputed missing values (IS not detected): 1205 (0.41%)\n",
      "\n",
      "MS2\n",
      "All values: 280140\n",
      "Values imputed with -5: 31885 (11.38%)\n",
      "Non-imputed missing values (IS not detected): 714 (0.25%)\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Count number of elements with -5 in MS2\n",
    "#\n",
    "\n",
    "logw('Missing Values')\n",
    "logw('')\n",
    "logw('MS1')\n",
    "logw(f\"All values: {xm1.shape[0]*xm1.shape[1]}\")\n",
    "logw(f\"Values imputed with -5: {(xm1==-5).sum().sum()} ({round(100*(xm1==-5).sum().sum()/(xm1.shape[0]*xm1.shape[1]),2)}%)\")\n",
    "logw(f'Non-imputed missing values (IS not detected): {xm1.isna().sum().sum()} ({round(100*xm1.isna().sum().sum()/(xm1.shape[0]*xm1.shape[1]),2)}%)')\n",
    "\n",
    "logw('')\n",
    "logw('MS2')\n",
    "logw(f\"All values: {xm2.shape[0]*xm2.shape[1]}\")\n",
    "logw(f\"Values imputed with -5: {(xm2==-5).sum().sum()} ({round(100*(xm2==-5).sum().sum()/(xm2.shape[0]*xm2.shape[1]),2)}%)\")\n",
    "logw(f'Non-imputed missing values (IS not detected): {xm2.isna().sum().sum()} ({round(100*xm2.isna().sum().sum()/(xm2.shape[0]*xm2.shape[1]),2)}%)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotMV = PlotMV(xm1, mdata, file=fileMSV)\n",
    "plotMV.plotSummary(titleLabel='- MS1')\n",
    "\n",
    "plotMV = PlotMV(xm2, mdata, file=fileMSV)\n",
    "plotMV.plotSummary(titleLabel='- MS2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Accepted features as a function of ratio of imputed values\n",
    "#\n",
    "\n",
    "xdata = np.arange(0,1,0.01)\n",
    "\n",
    "for df, i in [(xm1, 'MS1'), (xm2, 'MS2')]:\n",
    "\n",
    "    fig = go.Figure()\n",
    "    fig.add_trace(go.Scatter(\n",
    "        x=xdata,\n",
    "        y=[((df==-5).sum()/df.shape[0]<=i).sum() for i in xdata]\n",
    "    ))\n",
    "\n",
    "    fig.update_layout(\n",
    "        title=f'Accepted Features vs Imputed values {i}'\n",
    "    )\n",
    "\n",
    "    fig.update_xaxes(title='Ratio of imputed values')\n",
    "    fig.update_yaxes(title='Number of accepted features')\n",
    "    saveToFile(fig, fileMSV)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MS1\n",
      "Total number of observations: 348\n",
      "Total number of features: 849\n",
      "Total number of features with <20.0% of missing values(<69 of obs.): 849\n",
      "Total number of features with <20.0% of (-5)-imputed values (<69 of obs.): 773\n",
      "\n",
      "MS2\n",
      "Total number of observations: 348\n",
      "Total number of features: 805\n",
      "Total number of features with <20.0% of missing values(<69 of obs.): 805\n",
      "Total number of features with <20.0% of (-5)-imputed values (<69 of obs.): 656\n"
     ]
    }
   ],
   "source": [
    "logw('')\n",
    "logw('MS1')\n",
    "logw(f\"Total number of observations: {xm1.shape[0]}\")\n",
    "logw(f\"Total number of features: {xm1.shape[1]}\")\n",
    "logw(f\"Total number of features with <{MVF_thr*100}% of missing values(<{int(xm1.shape[0]*MVF_thr)} of obs.): {((xm1.isna().sum()/xm1.shape[0])<=MVF_thr).sum()}\")\n",
    "logw(f\"Total number of features with <{MVF_thr*100}% of (-5)-imputed values (<{int(xm1.shape[0]*MVF_thr)} of obs.): {(((xm1==-5).sum()/xm1.shape[0])<=MVF_thr).sum()}\")\n",
    "logw('')\n",
    "logw('MS2')\n",
    "logw(f\"Total number of observations: {xm2.shape[0]}\")\n",
    "logw(f\"Total number of features: {xm2.shape[1]}\")\n",
    "logw(f\"Total number of features with <{MVF_thr*100}% of missing values(<{int(xm2.shape[0]*MVF_thr)} of obs.): {((xm2.isna().sum()/xm2.shape[0])<=MVF_thr).sum()}\")\n",
    "logw(f\"Total number of features with <{MVF_thr*100}% of (-5)-imputed values (<{int(xm2.shape[0]*MVF_thr)} of obs.): {(((xm2==-5).sum()/xm2.shape[0])<=MVF_thr).sum()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Filter by missing values (although no feature will be removed)\n",
    "#\n",
    "\n",
    "xm1f = xm1.loc[:, xm1.isna().sum()/xm1.shape[0] <= MVF_thr]\n",
    "xm2f = xm2.loc[:, xm2.isna().sum()/xm2.shape[0] <= MVF_thr]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Remove values where >20% were (-5)-imputed\n",
    "#\n",
    "\n",
    "xm1f = xm1f.loc[:, (xm1f==-5).sum()/xm1f.shape[0]<MV5_thr]\n",
    "xm2f = xm2f.loc[:, (xm2f==-5).sum()/xm2f.shape[0]<MV5_thr]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>P1</th>\n",
       "      <th>P2</th>\n",
       "      <th>P4</th>\n",
       "      <th>P5</th>\n",
       "      <th>P6</th>\n",
       "      <th>P7</th>\n",
       "      <th>P8</th>\n",
       "      <th>P9</th>\n",
       "      <th>P10</th>\n",
       "      <th>P11</th>\n",
       "      <th>...</th>\n",
       "      <th>N174</th>\n",
       "      <th>N175</th>\n",
       "      <th>N176</th>\n",
       "      <th>N178</th>\n",
       "      <th>N179</th>\n",
       "      <th>N180</th>\n",
       "      <th>N181</th>\n",
       "      <th>N183</th>\n",
       "      <th>N188</th>\n",
       "      <th>N192</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2515</th>\n",
       "      <td>2.095463</td>\n",
       "      <td>1.092414</td>\n",
       "      <td>0.731614</td>\n",
       "      <td>2.004638</td>\n",
       "      <td>3.591444</td>\n",
       "      <td>1.601048</td>\n",
       "      <td>3.156607</td>\n",
       "      <td>1.812960</td>\n",
       "      <td>1.053862</td>\n",
       "      <td>1.125457</td>\n",
       "      <td>...</td>\n",
       "      <td>1.452542</td>\n",
       "      <td>2.475839</td>\n",
       "      <td>1.225159</td>\n",
       "      <td>2.074344</td>\n",
       "      <td>2.397792</td>\n",
       "      <td>2.224788</td>\n",
       "      <td>1.522018</td>\n",
       "      <td>0.619471</td>\n",
       "      <td>1.989512</td>\n",
       "      <td>2.589751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273710</th>\n",
       "      <td>1.943931</td>\n",
       "      <td>0.790746</td>\n",
       "      <td>0.593961</td>\n",
       "      <td>2.007067</td>\n",
       "      <td>3.395482</td>\n",
       "      <td>1.525246</td>\n",
       "      <td>3.083728</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-5.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.358770</td>\n",
       "      <td>2.417085</td>\n",
       "      <td>0.932336</td>\n",
       "      <td>1.898146</td>\n",
       "      <td>2.422405</td>\n",
       "      <td>2.084777</td>\n",
       "      <td>1.746221</td>\n",
       "      <td>-5.000000</td>\n",
       "      <td>1.646769</td>\n",
       "      <td>2.534357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353918</th>\n",
       "      <td>1.997492</td>\n",
       "      <td>0.717726</td>\n",
       "      <td>0.586638</td>\n",
       "      <td>1.822292</td>\n",
       "      <td>3.508992</td>\n",
       "      <td>1.755538</td>\n",
       "      <td>2.977030</td>\n",
       "      <td>1.916527</td>\n",
       "      <td>0.988410</td>\n",
       "      <td>1.555373</td>\n",
       "      <td>...</td>\n",
       "      <td>1.349207</td>\n",
       "      <td>2.392533</td>\n",
       "      <td>1.240281</td>\n",
       "      <td>2.010245</td>\n",
       "      <td>2.189663</td>\n",
       "      <td>1.945928</td>\n",
       "      <td>1.291314</td>\n",
       "      <td>0.436417</td>\n",
       "      <td>1.645139</td>\n",
       "      <td>2.536466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233292</th>\n",
       "      <td>1.824169</td>\n",
       "      <td>0.699200</td>\n",
       "      <td>0.345699</td>\n",
       "      <td>1.832666</td>\n",
       "      <td>3.510665</td>\n",
       "      <td>1.658584</td>\n",
       "      <td>3.069936</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.704044</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.398204</td>\n",
       "      <td>2.265840</td>\n",
       "      <td>1.275511</td>\n",
       "      <td>1.833980</td>\n",
       "      <td>2.212894</td>\n",
       "      <td>1.882090</td>\n",
       "      <td>1.392809</td>\n",
       "      <td>0.672716</td>\n",
       "      <td>1.544479</td>\n",
       "      <td>2.450986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456052</th>\n",
       "      <td>2.023524</td>\n",
       "      <td>1.210670</td>\n",
       "      <td>0.800484</td>\n",
       "      <td>1.914443</td>\n",
       "      <td>3.667758</td>\n",
       "      <td>1.882119</td>\n",
       "      <td>3.134274</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.365654</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.771895</td>\n",
       "      <td>2.437158</td>\n",
       "      <td>1.428423</td>\n",
       "      <td>1.929183</td>\n",
       "      <td>2.368912</td>\n",
       "      <td>2.101952</td>\n",
       "      <td>1.045453</td>\n",
       "      <td>1.167779</td>\n",
       "      <td>1.835953</td>\n",
       "      <td>2.467317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504335</th>\n",
       "      <td>1.910573</td>\n",
       "      <td>0.894368</td>\n",
       "      <td>0.689580</td>\n",
       "      <td>2.032865</td>\n",
       "      <td>3.623343</td>\n",
       "      <td>1.598363</td>\n",
       "      <td>3.108388</td>\n",
       "      <td>1.693454</td>\n",
       "      <td>0.857709</td>\n",
       "      <td>1.289752</td>\n",
       "      <td>...</td>\n",
       "      <td>1.615657</td>\n",
       "      <td>2.436585</td>\n",
       "      <td>1.349227</td>\n",
       "      <td>1.961006</td>\n",
       "      <td>2.297576</td>\n",
       "      <td>2.065189</td>\n",
       "      <td>1.035804</td>\n",
       "      <td>0.777547</td>\n",
       "      <td>1.765119</td>\n",
       "      <td>2.547947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>575679</th>\n",
       "      <td>2.125031</td>\n",
       "      <td>1.023011</td>\n",
       "      <td>0.803847</td>\n",
       "      <td>1.902651</td>\n",
       "      <td>3.526522</td>\n",
       "      <td>1.760763</td>\n",
       "      <td>3.169825</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.603663</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.727060</td>\n",
       "      <td>2.544028</td>\n",
       "      <td>1.510688</td>\n",
       "      <td>2.106131</td>\n",
       "      <td>2.363070</td>\n",
       "      <td>2.290222</td>\n",
       "      <td>1.134026</td>\n",
       "      <td>0.938151</td>\n",
       "      <td>1.979531</td>\n",
       "      <td>2.657189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>733797</th>\n",
       "      <td>2.126363</td>\n",
       "      <td>0.950849</td>\n",
       "      <td>0.779019</td>\n",
       "      <td>2.079231</td>\n",
       "      <td>3.541726</td>\n",
       "      <td>1.619899</td>\n",
       "      <td>3.135876</td>\n",
       "      <td>1.758719</td>\n",
       "      <td>0.961915</td>\n",
       "      <td>1.357612</td>\n",
       "      <td>...</td>\n",
       "      <td>1.584028</td>\n",
       "      <td>2.528782</td>\n",
       "      <td>1.420058</td>\n",
       "      <td>2.111025</td>\n",
       "      <td>2.338601</td>\n",
       "      <td>2.081349</td>\n",
       "      <td>0.995462</td>\n",
       "      <td>0.763925</td>\n",
       "      <td>1.838135</td>\n",
       "      <td>2.668286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>735630</th>\n",
       "      <td>2.035358</td>\n",
       "      <td>0.289307</td>\n",
       "      <td>0.508292</td>\n",
       "      <td>1.845639</td>\n",
       "      <td>3.467188</td>\n",
       "      <td>1.617190</td>\n",
       "      <td>3.109096</td>\n",
       "      <td>1.790774</td>\n",
       "      <td>1.049544</td>\n",
       "      <td>1.318179</td>\n",
       "      <td>...</td>\n",
       "      <td>1.543145</td>\n",
       "      <td>2.465849</td>\n",
       "      <td>1.380361</td>\n",
       "      <td>2.014429</td>\n",
       "      <td>2.269131</td>\n",
       "      <td>2.138384</td>\n",
       "      <td>0.899434</td>\n",
       "      <td>0.301049</td>\n",
       "      <td>1.784816</td>\n",
       "      <td>2.655234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>922936</th>\n",
       "      <td>1.658357</td>\n",
       "      <td>0.652635</td>\n",
       "      <td>0.485378</td>\n",
       "      <td>1.476683</td>\n",
       "      <td>3.365485</td>\n",
       "      <td>1.719752</td>\n",
       "      <td>2.816319</td>\n",
       "      <td>1.650729</td>\n",
       "      <td>1.024456</td>\n",
       "      <td>1.209205</td>\n",
       "      <td>...</td>\n",
       "      <td>1.320304</td>\n",
       "      <td>2.221254</td>\n",
       "      <td>1.229221</td>\n",
       "      <td>1.804539</td>\n",
       "      <td>2.015647</td>\n",
       "      <td>1.832525</td>\n",
       "      <td>1.125449</td>\n",
       "      <td>0.364763</td>\n",
       "      <td>1.405122</td>\n",
       "      <td>2.405463</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>348 rows × 773 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              P1        P2        P4        P5        P6        P7        P8  \\\n",
       "fid                                                                            \n",
       "2515    2.095463  1.092414  0.731614  2.004638  3.591444  1.601048  3.156607   \n",
       "273710  1.943931  0.790746  0.593961  2.007067  3.395482  1.525246  3.083728   \n",
       "353918  1.997492  0.717726  0.586638  1.822292  3.508992  1.755538  2.977030   \n",
       "233292  1.824169  0.699200  0.345699  1.832666  3.510665  1.658584  3.069936   \n",
       "456052  2.023524  1.210670  0.800484  1.914443  3.667758  1.882119  3.134274   \n",
       "...          ...       ...       ...       ...       ...       ...       ...   \n",
       "504335  1.910573  0.894368  0.689580  2.032865  3.623343  1.598363  3.108388   \n",
       "575679  2.125031  1.023011  0.803847  1.902651  3.526522  1.760763  3.169825   \n",
       "733797  2.126363  0.950849  0.779019  2.079231  3.541726  1.619899  3.135876   \n",
       "735630  2.035358  0.289307  0.508292  1.845639  3.467188  1.617190  3.109096   \n",
       "922936  1.658357  0.652635  0.485378  1.476683  3.365485  1.719752  2.816319   \n",
       "\n",
       "              P9       P10       P11  ...      N174      N175      N176  \\\n",
       "fid                                   ...                                 \n",
       "2515    1.812960  1.053862  1.125457  ...  1.452542  2.475839  1.225159   \n",
       "273710       NaN -5.000000       NaN  ...  1.358770  2.417085  0.932336   \n",
       "353918  1.916527  0.988410  1.555373  ...  1.349207  2.392533  1.240281   \n",
       "233292       NaN  0.704044       NaN  ...  1.398204  2.265840  1.275511   \n",
       "456052       NaN  1.365654       NaN  ...  1.771895  2.437158  1.428423   \n",
       "...          ...       ...       ...  ...       ...       ...       ...   \n",
       "504335  1.693454  0.857709  1.289752  ...  1.615657  2.436585  1.349227   \n",
       "575679       NaN  0.603663       NaN  ...  1.727060  2.544028  1.510688   \n",
       "733797  1.758719  0.961915  1.357612  ...  1.584028  2.528782  1.420058   \n",
       "735630  1.790774  1.049544  1.318179  ...  1.543145  2.465849  1.380361   \n",
       "922936  1.650729  1.024456  1.209205  ...  1.320304  2.221254  1.229221   \n",
       "\n",
       "            N178      N179      N180      N181      N183      N188      N192  \n",
       "fid                                                                           \n",
       "2515    2.074344  2.397792  2.224788  1.522018  0.619471  1.989512  2.589751  \n",
       "273710  1.898146  2.422405  2.084777  1.746221 -5.000000  1.646769  2.534357  \n",
       "353918  2.010245  2.189663  1.945928  1.291314  0.436417  1.645139  2.536466  \n",
       "233292  1.833980  2.212894  1.882090  1.392809  0.672716  1.544479  2.450986  \n",
       "456052  1.929183  2.368912  2.101952  1.045453  1.167779  1.835953  2.467317  \n",
       "...          ...       ...       ...       ...       ...       ...       ...  \n",
       "504335  1.961006  2.297576  2.065189  1.035804  0.777547  1.765119  2.547947  \n",
       "575679  2.106131  2.363070  2.290222  1.134026  0.938151  1.979531  2.657189  \n",
       "733797  2.111025  2.338601  2.081349  0.995462  0.763925  1.838135  2.668286  \n",
       "735630  2.014429  2.269131  2.138384  0.899434  0.301049  1.784816  2.655234  \n",
       "922936  1.804539  2.015647  1.832525  1.125449  0.364763  1.405122  2.405463  \n",
       "\n",
       "[348 rows x 773 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xm1f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of observations with <10.0% of missing values: 347 / 348\n",
      "Total number of observations with <10.0% of missing values: 348 / 348\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Filter observations by Missing values\n",
    "#\n",
    "\n",
    "# MS1\n",
    "plotMV = PlotMV(xm1f, mdata, file=fileMSV)\n",
    "plotMV.plotSummaryObs()\n",
    "\n",
    "# Filter Observations by missing values\n",
    "\n",
    "xm1f = xm1f[xm1f.isna().sum(axis=1)/xm1f.shape[1]<MVO_thr]\n",
    "\n",
    "logw(f'Total number of observations with <{MVO_thr*100}% of missing values: {xm1f.shape[0]} / {xm1.shape[0]}')\n",
    "\n",
    "\n",
    "# MS2\n",
    "plotMV = PlotMV(xm2f, mdata, file=fileMSV)\n",
    "plotMV.plotSummaryObs()\n",
    "\n",
    "# Filter Observations by missing values\n",
    "\n",
    "xm2f = xm2f[xm2f.isna().sum(axis=1)/xm2f.shape[1]<MVO_thr]\n",
    "\n",
    "logw(f'Total number of observations with <{MVO_thr*100}% of missing values: {xm2f.shape[0]} / {xm2.shape[0]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Summary Plots\n",
    "# Probamos dos tipos de estandarizacion\n",
    "\n",
    "# MS1\n",
    "for i in ['s', 'r']: # StandardScaler and RobustScaler\n",
    "    plotEDA = PlotEDA(xm1f, mdata, file=fileSummaryMS1, scaler=i)\n",
    "    plotEDA.plotSummary(\n",
    "        r11=(-5, 5), r12=(0,3), r3=(-6,6),\n",
    "        vl3=[0],\n",
    "        binsize=0.1,\n",
    "        titleLabel=f'- Scaler: {\"Standard\" if i==\"s\" else \"Robust\"}'\n",
    "    )\n",
    "    plotEDA.plotByGroup('Metabo_Batch',vl1=[0],vl2=[0], r1=(-6,6), r2=(-6,6))\n",
    "    plotEDA.plotByGroup('Group',vl1=[0],vl2=[0], r1=(-6,6), r2=(-6,6))\n",
    "\n",
    "\n",
    "# MS2\n",
    "for i,file in zip(['s', 'r'],[fileSummaryMS2S,fileSummaryMS2R]):\n",
    "    plotEDA = PlotEDA(xm2f, mdata, file=file, scaler=i)\n",
    "    plotEDA.plotSummary(\n",
    "        r11=(-5, 5), r12=(0,3), r3=(-6,6),\n",
    "        vl3=[0],\n",
    "        binsize=0.1,\n",
    "        titleLabel=f'- Scaler: {\"Standard\" if i==\"s\" else \"Robust\"}'\n",
    "    )\n",
    "    plotEDA.plotByGroup('Metabo_Batch',vl1=[0],vl2=[0], r1=(-6,6), r2=(-6,6))\n",
    "    plotEDA.plotByGroup('Group',vl1=[0],vl2=[0], r1=(-6,6), r2=(-6,6))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Standardize\n",
    "#\n",
    "\n",
    "from sklearn.preprocessing import StandardScaler, RobustScaler\n",
    "\n",
    "xm1fn = pd.DataFrame(\n",
    "    StandardScaler().fit_transform(xm1f),\n",
    "    columns=xm1f.columns, index=xm1f.index\n",
    ")\n",
    "\n",
    "xm2fn = pd.DataFrame(\n",
    "    StandardScaler().fit_transform(xm2f),\n",
    "    columns=xm2f.columns, index=xm2f.index\n",
    ")\n",
    "\n",
    "xm1fr = pd.DataFrame(\n",
    "    RobustScaler().fit_transform(xm1f),\n",
    "    columns=xm1f.columns, index=xm1f.index\n",
    ")\n",
    "\n",
    "xm2fr = pd.DataFrame(\n",
    "    RobustScaler().fit_transform(xm2f),\n",
    "    columns=xm2f.columns, index=xm2f.index\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Imputation of missing values using KNN\n",
    "#\n",
    "\n",
    "from sklearn.impute import KNNImputer\n",
    "\n",
    "\n",
    "xm1fnv = pd.DataFrame(\n",
    "    KNNImputer(n_neighbors=3).fit_transform(xm1fn),\n",
    "    columns=xm1fn.columns,\n",
    "    index=xm1fn.index\n",
    ")\n",
    "\n",
    "xm2fnv = pd.DataFrame(\n",
    "    KNNImputer(n_neighbors=3).fit_transform(xm2fn),\n",
    "    columns=xm2fn.columns,\n",
    "    index=xm2fn.index\n",
    ")\n",
    "\n",
    "xm1frv = pd.DataFrame(\n",
    "    KNNImputer(n_neighbors=3).fit_transform(xm1fr),\n",
    "    columns=xm1fr.columns,\n",
    "    index=xm1fr.index\n",
    ")\n",
    "\n",
    "xm2frv = pd.DataFrame(\n",
    "    KNNImputer(n_neighbors=3).fit_transform(xm2fr),\n",
    "    columns=xm2fr.columns,\n",
    "    index=xm2fr.index\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Imputed missing values in MS1:     | 'KNN' 1074/268231 (0.4%)     | '-5' 3328/268231 (1.24%) \n",
      "Imputed missing values in MS2:     | 'KNN' 572/228288 (0.25%)     | '-5' 6373/228288 (2.79%) \n"
     ]
    }
   ],
   "source": [
    "logw(f\"Imputed missing values in MS1: \\\n",
    "    | 'KNN' {xm1f.isna().sum().sum()}/{xm1f.shape[0]*xm1f.shape[1]} ({round(xm1f.isna().sum().sum()/(xm1f.shape[0]*xm1f.shape[1])*100, 2)}%) \\\n",
    "    | '-5' {(xm1f==-5).sum().sum()}/{xm1f.shape[0]*xm1f.shape[1]} ({round((xm1f==-5).sum().sum()/(xm1f.shape[0]*xm1f.shape[1])*100, 2)}%) \\\n",
    "\")\n",
    "logw(f\"Imputed missing values in MS2: \\\n",
    "    | 'KNN' {xm2f.isna().sum().sum()}/{xm2f.shape[0]*xm2f.shape[1]} ({round(xm2f.isna().sum().sum()/(xm2f.shape[0]*xm2f.shape[1])*100, 2)}%) \\\n",
    "    | '-5' {(xm2f==-5).sum().sum()}/{xm2f.shape[0]*xm2f.shape[1]} ({round((xm2f==-5).sum().sum()/(xm2f.shape[0]*xm2f.shape[1])*100, 2)}%) \\\n",
    "\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading required package: mgcv\n",
      "Loading required package: nlme\n",
      "This is mgcv 1.8-41. For overview type 'help(\"mgcv-package\")'.\n",
      "Loading required package: genefilter\n",
      "Loading required package: BiocParallel\n",
      "Warning messages:\n",
      "1: package 'sva' was built under R version 4.2.1 \n",
      "2: package 'mgcv' was built under R version 4.2.2 \n",
      "3: package 'nlme' was built under R version 4.2.2 \n",
      "4: package 'genefilter' was built under R version 4.2.2 \n",
      "5: package 'BiocParallel' was built under R version 4.2.2 \n",
      "Found8batches\n",
      "Adjusting for7covariate(s) or covariate level(s)\n",
      "Standardizing Data across genes\n",
      "Fitting L/S model and finding priors\n",
      "Finding nonparametric adjustments\n",
      "Adjusting the Data\n",
      "\n",
      "\n",
      "Loading required package: mgcv\n",
      "Loading required package: nlme\n",
      "This is mgcv 1.8-41. For overview type 'help(\"mgcv-package\")'.\n",
      "Loading required package: genefilter\n",
      "Loading required package: BiocParallel\n",
      "Warning messages:\n",
      "1: package 'sva' was built under R version 4.2.1 \n",
      "2: package 'mgcv' was built under R version 4.2.2 \n",
      "3: package 'nlme' was built under R version 4.2.2 \n",
      "4: package 'genefilter' was built under R version 4.2.2 \n",
      "5: package 'BiocParallel' was built under R version 4.2.2 \n",
      "Found8batches\n",
      "Adjusting for7covariate(s) or covariate level(s)\n",
      "Standardizing Data across genes\n",
      "Fitting L/S model and finding priors\n",
      "Finding nonparametric adjustments\n",
      "Adjusting the Data\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Correct Batch Effect\n",
    "# Comparamos batch effect correction con los dos tipos de escalado\n",
    "# https://github.com/brentp/combat.py\n",
    "\n",
    "# from combat import combat\n",
    "# from scipy.stats import kruskal, median_test\n",
    "\n",
    "# xm2fnvb = combat(\n",
    "#     data=xm2fnv.T,\n",
    "#     batch=mdata.set_index('Seqn').loc[xm2fnv.index, 'Cohort']\n",
    "# ).T\n",
    "\n",
    "# xm2frvb = combat(\n",
    "#     data=xm2frv.T,\n",
    "#     batch=mdata.set_index('Seqn').loc[xm2frv.index, 'Cohort']\n",
    "# ).T\n",
    "\n",
    "#\n",
    "# Apply myComBat\n",
    "# We take R ComBat from sva package\n",
    "#\n",
    "\n",
    "from myComBat import myComBat\n",
    "\n",
    "catVars = ['Group', 'smoker', 'diabetes']\n",
    "conVars = ['Plaque thickness', 'Glucosa', 'hdl', 'coltot']\n",
    "xm2fnvb = myComBat(xm2fnv, mdata, 'Metabo_Batch', catVars, conVars, \n",
    "                   Rpath=os.path.join(workingPath, 'WorkingFiles', 'myRData'),\n",
    "                    )#Rengine=r\"C:\\Program Files\\R\\R-4.1.2\\bin\\Rscript.exe\")\n",
    "\n",
    "xm2frvb = myComBat(xm2frv, mdata, 'Metabo_Batch', catVars, conVars, \n",
    "                   Rpath=os.path.join(workingPath, 'WorkingFiles', 'myRData'),\n",
    "                    )#Rengine=r\"C:\\Program Files\\R\\R-4.1.2\\bin\\Rscript.exe\")\n",
    "\n",
    "ploteda = PlotEDA(xm2fnvb, mdata, file=fileSummaryMS2S)\n",
    "ploteda.plotByGroup('Metabo_Batch', plotN=False, titleLabel='- Batch Corrected (Standard)')\n",
    "\n",
    "ploteda = PlotEDA(xm2frvb, mdata, file=fileSummaryMS2R)\n",
    "ploteda.plotByGroup('Metabo_Batch', plotN=False, titleLabel='- Batch Corrected (Robust)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "ploteda = PlotEDA(xm2fnvb, mdata, file=fileSummaryMS2S)\n",
    "ploteda.plotByGroup('Group', plotN=False, titleLabel='- Batch Corrected (Standard)')\n",
    "\n",
    "ploteda = PlotEDA(xm2frvb, mdata, file=fileSummaryMS2R)\n",
    "ploteda.plotByGroup('Group', plotN=False, titleLabel='- Batch Corrected (Robust)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kruskal-Wallis:\n",
      "KruskalResult(statistic=87.63569094818578, pvalue=3.7810369109996196e-16)\n",
      "PostHoc-MannWhitney\n",
      "          1             2             3             4         5             6  \\\n",
      "1  1.000000  1.000000e+00  1.000000e+00  1.000000e+00  0.001193  1.064335e-06   \n",
      "2  1.000000  1.000000e+00  1.000000e+00  1.000000e+00  0.000041  7.799078e-09   \n",
      "3  1.000000  1.000000e+00  1.000000e+00  1.000000e+00  0.000161  6.944119e-08   \n",
      "4  1.000000  1.000000e+00  1.000000e+00  1.000000e+00  0.000515  6.576818e-07   \n",
      "5  0.001193  4.130242e-05  1.612383e-04  5.148815e-04  1.000000  1.000000e+00   \n",
      "6  0.000001  7.799078e-09  6.944119e-08  6.576818e-07  1.000000  1.000000e+00   \n",
      "7  0.000265  1.216554e-04  1.948148e-04  1.096842e-03  1.000000  1.000000e+00   \n",
      "8  1.000000  1.000000e+00  1.000000e+00  1.000000e+00  0.015944  1.236629e-05   \n",
      "\n",
      "          7         8  \n",
      "1  0.000265  1.000000  \n",
      "2  0.000122  1.000000  \n",
      "3  0.000195  1.000000  \n",
      "4  0.001097  1.000000  \n",
      "5  1.000000  0.015944  \n",
      "6  1.000000  0.000012  \n",
      "7  1.000000  0.023275  \n",
      "8  0.023275  1.000000  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "KruskalResult(statistic=87.63569094818578, pvalue=3.7810369109996196e-16)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PlotEDA(xm2fnvb, mdata)._kruskal(xm2fnvb, 'Metabo_Batch', showTest=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Generate \n",
    "#\n",
    "\n",
    "xm1fnv.to_csv(os.path.join(workingPath, 'WorkingFiles', 'Xm_norm_MS1.tsv'), sep='\\t')\n",
    "xm2fnvb.to_csv(os.path.join(workingPath, 'WorkingFiles', 'Xm_norm_MS2.tsv'), sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Dimensionality Reduction\n",
    "#\n",
    "\n",
    "pcaumap = PCA_UMAP(xm1fnv, mdata, file=filePCAMS1)\n",
    "pcaumap.plotReduction('Metabo_Batch', pcacomp=[0,1], titleLabel='Scaler: Standard')\n",
    "pcaumap.plotReduction('Group', titleLabel='Scaler: Standard')\n",
    "\n",
    "pcaumap = PCA_UMAP(xm2fnv, mdata, file=filePCAMS2)\n",
    "pcaumap.plotReduction('Metabo_Batch', pcacomp=[0,1], titleLabel='Scaler: Standard')\n",
    "pcaumap.plotReduction('Group', titleLabel='Scaler: Standard')\n",
    "\n",
    "pcaumap = PCA_UMAP(xm2fnvb, mdata, file=filePCAMS2)\n",
    "pcaumap.plotReduction('Metabo_Batch', pcacomp=[0,1], titleLabel='Scaler: Standard - Batch')\n",
    "pcaumap.plotReduction('Group', pcacomp=[0,1], titleLabel='Scaler: Standard - Batch')\n",
    "\n",
    "\n",
    "pcaumap = PCA_UMAP(xm1frv, mdata, file=filePCAMS1)\n",
    "pcaumap.plotReduction('Metabo_Batch', pcacomp=[0,1], titleLabel='Scaler: Robust')\n",
    "pcaumap.plotReduction('Group', titleLabel='Scaler: Robust')\n",
    "\n",
    "pcaumap = PCA_UMAP(xm2frv, mdata, file=filePCAMS2)\n",
    "pcaumap.plotReduction('Metabo_Batch', pcacomp=[0,1], titleLabel='Scaler: Robust')\n",
    "pcaumap.plotReduction('Group', titleLabel='Scaler: Robust')\n",
    "\n",
    "pcaumap = PCA_UMAP(xm2frvb, mdata, file=filePCAMS2)\n",
    "pcaumap.plotReduction('Metabo_Batch', pcacomp=[0,1], titleLabel='Scaler: Robust - Batch')\n",
    "pcaumap.plotReduction('Group', pcacomp=[0,1], titleLabel='Scaler: Robust - Batch')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>%Var PCA</th>\n",
       "      <th>Plaque thickness</th>\n",
       "      <th>Glucosa</th>\n",
       "      <th>hdl</th>\n",
       "      <th>coltot</th>\n",
       "      <th>Group</th>\n",
       "      <th>smoker</th>\n",
       "      <th>diabetes</th>\n",
       "      <th>Metabo_Batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30.636145</td>\n",
       "      <td>0.7239</td>\n",
       "      <td>0.0253</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0284</td>\n",
       "      <td>0.1331</td>\n",
       "      <td>0.1044</td>\n",
       "      <td>0.0158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6.645477</td>\n",
       "      <td>0.6940</td>\n",
       "      <td>0.5938</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0030</td>\n",
       "      <td>0.2594</td>\n",
       "      <td>0.2793</td>\n",
       "      <td>0.6969</td>\n",
       "      <td>0.0245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5.335242</td>\n",
       "      <td>0.4498</td>\n",
       "      <td>0.4286</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5372</td>\n",
       "      <td>0.3670</td>\n",
       "      <td>0.0890</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3.137564</td>\n",
       "      <td>0.4442</td>\n",
       "      <td>0.8364</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0217</td>\n",
       "      <td>0.8772</td>\n",
       "      <td>0.0020</td>\n",
       "      <td>0.4167</td>\n",
       "      <td>0.0112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2.976598</td>\n",
       "      <td>0.6568</td>\n",
       "      <td>0.3466</td>\n",
       "      <td>0.7924</td>\n",
       "      <td>0.3097</td>\n",
       "      <td>0.6445</td>\n",
       "      <td>0.0920</td>\n",
       "      <td>0.9944</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2.835709</td>\n",
       "      <td>0.3727</td>\n",
       "      <td>0.1757</td>\n",
       "      <td>0.8963</td>\n",
       "      <td>0.8449</td>\n",
       "      <td>0.8769</td>\n",
       "      <td>0.8090</td>\n",
       "      <td>0.4326</td>\n",
       "      <td>0.0002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2.296204</td>\n",
       "      <td>0.5788</td>\n",
       "      <td>0.5367</td>\n",
       "      <td>0.3941</td>\n",
       "      <td>0.0052</td>\n",
       "      <td>0.7837</td>\n",
       "      <td>0.8005</td>\n",
       "      <td>0.5992</td>\n",
       "      <td>0.0002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1.812636</td>\n",
       "      <td>0.4416</td>\n",
       "      <td>0.5748</td>\n",
       "      <td>0.9727</td>\n",
       "      <td>0.3193</td>\n",
       "      <td>0.8047</td>\n",
       "      <td>0.0039</td>\n",
       "      <td>0.5216</td>\n",
       "      <td>0.3760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1.717879</td>\n",
       "      <td>0.2527</td>\n",
       "      <td>0.1417</td>\n",
       "      <td>0.6077</td>\n",
       "      <td>0.8766</td>\n",
       "      <td>0.8722</td>\n",
       "      <td>0.6031</td>\n",
       "      <td>0.4029</td>\n",
       "      <td>0.0054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1.684994</td>\n",
       "      <td>0.2504</td>\n",
       "      <td>0.4444</td>\n",
       "      <td>0.4912</td>\n",
       "      <td>0.8829</td>\n",
       "      <td>0.8462</td>\n",
       "      <td>0.0835</td>\n",
       "      <td>0.5063</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     %Var PCA  Plaque thickness  Glucosa     hdl  coltot   Group  smoker  \\\n",
       "1   30.636145            0.7239   0.0253  0.0000  0.0000  0.0284  0.1331   \n",
       "2    6.645477            0.6940   0.5938  0.0000  0.0030  0.2594  0.2793   \n",
       "3    5.335242            0.4498   0.4286  0.0000  0.0000  0.5372  0.3670   \n",
       "4    3.137564            0.4442   0.8364  0.0013  0.0217  0.8772  0.0020   \n",
       "5    2.976598            0.6568   0.3466  0.7924  0.3097  0.6445  0.0920   \n",
       "6    2.835709            0.3727   0.1757  0.8963  0.8449  0.8769  0.8090   \n",
       "7    2.296204            0.5788   0.5367  0.3941  0.0052  0.7837  0.8005   \n",
       "8    1.812636            0.4416   0.5748  0.9727  0.3193  0.8047  0.0039   \n",
       "9    1.717879            0.2527   0.1417  0.6077  0.8766  0.8722  0.6031   \n",
       "10   1.684994            0.2504   0.4444  0.4912  0.8829  0.8462  0.0835   \n",
       "\n",
       "    diabetes  Metabo_Batch  \n",
       "1     0.1044        0.0158  \n",
       "2     0.6969        0.0245  \n",
       "3     0.0890        0.0001  \n",
       "4     0.4167        0.0112  \n",
       "5     0.9944        0.0000  \n",
       "6     0.4326        0.0002  \n",
       "7     0.5992        0.0002  \n",
       "8     0.5216        0.3760  \n",
       "9     0.4029        0.0054  \n",
       "10    0.5063        0.0000  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCA_Var(xm2fnv, mdata, conVars, catVars+['Metabo_Batch'], n_comp=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>%Var PCA</th>\n",
       "      <th>Plaque thickness</th>\n",
       "      <th>Glucosa</th>\n",
       "      <th>hdl</th>\n",
       "      <th>coltot</th>\n",
       "      <th>Group</th>\n",
       "      <th>smoker</th>\n",
       "      <th>diabetes</th>\n",
       "      <th>Metabo_Batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>33.492661</td>\n",
       "      <td>0.4134</td>\n",
       "      <td>0.0341</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0653</td>\n",
       "      <td>0.5422</td>\n",
       "      <td>0.1370</td>\n",
       "      <td>0.9888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7.681274</td>\n",
       "      <td>0.6661</td>\n",
       "      <td>0.9221</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0112</td>\n",
       "      <td>0.2237</td>\n",
       "      <td>0.9991</td>\n",
       "      <td>0.9691</td>\n",
       "      <td>0.9485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6.310365</td>\n",
       "      <td>0.5487</td>\n",
       "      <td>0.3287</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.6962</td>\n",
       "      <td>0.1845</td>\n",
       "      <td>0.0878</td>\n",
       "      <td>0.6678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3.449286</td>\n",
       "      <td>0.4217</td>\n",
       "      <td>0.2998</td>\n",
       "      <td>0.1011</td>\n",
       "      <td>0.0076</td>\n",
       "      <td>0.7965</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>0.9440</td>\n",
       "      <td>0.9783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3.066355</td>\n",
       "      <td>0.6041</td>\n",
       "      <td>0.8524</td>\n",
       "      <td>0.0480</td>\n",
       "      <td>0.1724</td>\n",
       "      <td>0.6836</td>\n",
       "      <td>0.3804</td>\n",
       "      <td>0.3949</td>\n",
       "      <td>0.9683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2.305752</td>\n",
       "      <td>0.7606</td>\n",
       "      <td>0.9650</td>\n",
       "      <td>0.9063</td>\n",
       "      <td>0.5670</td>\n",
       "      <td>0.3431</td>\n",
       "      <td>0.0081</td>\n",
       "      <td>0.8387</td>\n",
       "      <td>0.7378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1.791799</td>\n",
       "      <td>0.7388</td>\n",
       "      <td>0.3435</td>\n",
       "      <td>0.5400</td>\n",
       "      <td>0.6949</td>\n",
       "      <td>0.2115</td>\n",
       "      <td>0.4406</td>\n",
       "      <td>0.5209</td>\n",
       "      <td>0.9996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1.743469</td>\n",
       "      <td>0.3627</td>\n",
       "      <td>0.6445</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>0.6426</td>\n",
       "      <td>0.5901</td>\n",
       "      <td>0.6888</td>\n",
       "      <td>0.4672</td>\n",
       "      <td>0.9944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1.597932</td>\n",
       "      <td>0.4211</td>\n",
       "      <td>0.2412</td>\n",
       "      <td>0.0021</td>\n",
       "      <td>0.0948</td>\n",
       "      <td>0.6484</td>\n",
       "      <td>0.6435</td>\n",
       "      <td>0.9543</td>\n",
       "      <td>0.9832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1.447864</td>\n",
       "      <td>0.8043</td>\n",
       "      <td>0.4963</td>\n",
       "      <td>0.1382</td>\n",
       "      <td>0.0356</td>\n",
       "      <td>0.1699</td>\n",
       "      <td>0.6249</td>\n",
       "      <td>0.4248</td>\n",
       "      <td>0.9949</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     %Var PCA  Plaque thickness  Glucosa     hdl  coltot   Group  smoker  \\\n",
       "1   33.492661            0.4134   0.0341  0.0000  0.0000  0.0653  0.5422   \n",
       "2    7.681274            0.6661   0.9221  0.0000  0.0112  0.2237  0.9991   \n",
       "3    6.310365            0.5487   0.3287  0.0000  0.0001  0.6962  0.1845   \n",
       "4    3.449286            0.4217   0.2998  0.1011  0.0076  0.7965  0.0004   \n",
       "5    3.066355            0.6041   0.8524  0.0480  0.1724  0.6836  0.3804   \n",
       "6    2.305752            0.7606   0.9650  0.9063  0.5670  0.3431  0.0081   \n",
       "7    1.791799            0.7388   0.3435  0.5400  0.6949  0.2115  0.4406   \n",
       "8    1.743469            0.3627   0.6445  0.0004  0.6426  0.5901  0.6888   \n",
       "9    1.597932            0.4211   0.2412  0.0021  0.0948  0.6484  0.6435   \n",
       "10   1.447864            0.8043   0.4963  0.1382  0.0356  0.1699  0.6249   \n",
       "\n",
       "    diabetes  Metabo_Batch  \n",
       "1     0.1370        0.9888  \n",
       "2     0.9691        0.9485  \n",
       "3     0.0878        0.6678  \n",
       "4     0.9440        0.9783  \n",
       "5     0.3949        0.9683  \n",
       "6     0.8387        0.7378  \n",
       "7     0.5209        0.9996  \n",
       "8     0.4672        0.9944  \n",
       "9     0.9543        0.9832  \n",
       "10    0.4248        0.9949  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCA_Var(xm2fnvb, mdata, conVars, catVars+['Metabo_Batch'], n_comp=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Group  Metabo_Batch\n",
       "C      1.0             22\n",
       "       2.0             22\n",
       "       3.0             22\n",
       "       4.0             22\n",
       "       5.0             22\n",
       "       6.0             22\n",
       "       7.0             22\n",
       "       8.0             20\n",
       "D      1.0             22\n",
       "       2.0             22\n",
       "       3.0             22\n",
       "       4.0             22\n",
       "       5.0             22\n",
       "       6.0             22\n",
       "       7.0             22\n",
       "       8.0             20\n",
       "dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdatam.loc[:, ['Group', 'Metabo_Batch']].groupby(['Group', 'Metabo_Batch']).size()\n",
    "\n",
    "#mdata['Cohort'].value_counts()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "299fabdee10379681b2207a83aa9f93c313ee5d5504e286e44b436bd7e45d8f9"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
