{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Import Libraries\n",
    "#\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "from functools import reduce\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "#\n",
    "# Import own libraries\n",
    "#\n",
    "utilsPath = r'S:\\U_Proteomica\\UNIDAD\\software\\MacrosRafa\\data\\Metabolomics\\PESA_Integromics\\Data\\utils'\n",
    "if utilsPath not in sys.path:\n",
    "    sys.path.append(utilsPath)\n",
    "\n",
    "from myLog import myLog\n",
    "from PlotEDA import PlotEDA\n",
    "from PlotMV import PlotMV\n",
    "from PCA_UMAP import PCA_UMAP, PCA_Var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Set constants\n",
    "#\n",
    "\n",
    "working_path = r'S:\\U_Proteomica\\UNIDAD\\software\\MacrosRafa\\data\\Metabolomics\\PESA_Integromics\\Data\\Proteomics\\ALDH4'\n",
    "n2i_path = os.path.join(working_path, 'OriginalFiles', 'RBR_n2info.tsv')\n",
    "\n",
    "fileSummary = os.path.join(working_path, 'WorkingFiles', 'Plots', 'SummaryPlots.html')\n",
    "filePCA = os.path.join(working_path, 'WorkingFiles', 'Plots', 'PCAPlots.html')\n",
    "if os.path.exists(fileSummary): os.remove(fileSummary)\n",
    "if os.path.exists(filePCA): os.remove(filePCA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start Session\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Set logging\n",
    "#\n",
    "\n",
    "logw = myLog(os.path.join(working_path, 'WorkingFiles', 'info.log'))\n",
    "logw('Start Session')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Read Experiment information\n",
    "#\n",
    "\n",
    "n2i = pd.read_csv(n2i_path, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Read X' values\n",
    "#\n",
    "\n",
    "xq = [\n",
    "    pd.read_csv(f, sep='\\t')\n",
    "        .rename(columns={\"X'inf\": f'{m}-{t}'})\n",
    "        .drop('Winf', axis=1)\n",
    "        #.set_index('idinf')\n",
    "    for m, f, t in zip(n2i['Mouse'], n2i['file'], n2i['TMT'])\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "xq = [\n",
    "    df.set_index(\n",
    "        pd.Index([re.search(r'\\|([^|]+)\\|', i).groups()[0] for i in df['idinf']])\n",
    "    ) \n",
    "    for df in xq\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Generate q2info.tsv file\n",
    "#\n",
    "\n",
    "q2i = pd.DataFrame(\n",
    "    [i for df in xq for i in zip(df.index.to_list(), df['idinf'])],\n",
    "    columns=['fid', 'qdesc']\n",
    ").drop_duplicates()\n",
    "\n",
    "idq1 = pd.read_csv(r\"S:\\U_Proteomica\\LABS\\LAB_ARR\\ClonesAb-atherosclerosis\\Higados\\proteomica\\TMT1\\msf\\ID_Q_XV.txt\", sep='\\t')\n",
    "idq2 = pd.read_csv(r\"S:\\U_Proteomica\\LABS\\LAB_ARR\\ClonesAb-atherosclerosis\\Higados\\proteomica\\TMT2\\msf\\ID_Q_XV.txt\", sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nps(idq, suffix):\n",
    "\n",
    "    nps = pd.DataFrame([\n",
    "        (\n",
    "            re.search(r'\\|([^|]+)\\|', i).groups()[0],\n",
    "            ''.join(re.findall(r'[A-Z]+', j))\n",
    "        )\n",
    "        for i,j in zip(idq['FASTAProteinDescription'].to_list(), idq['Sequence'].to_list())\n",
    "    ], columns=['fid', 'seq'])\n",
    "\n",
    "    nps = nps['fid'].value_counts().to_frame().rename(columns={'fid': f'ScanFreq_{suffix}'}).join(\n",
    "        nps.drop_duplicates()['fid'].value_counts().to_frame().rename(columns={'fid': f'pFreq_{suffix}'}),\n",
    "        how='outer'\n",
    "    )\n",
    "\n",
    "    return nps\n",
    "\n",
    "\n",
    "nps = get_nps(idq1, 1).join(\n",
    "    get_nps(idq2, 2),\n",
    "    how='outer'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "q2i = q2i.set_index('fid').join(nps)\n",
    "q2i.to_csv(os.path.join(working_path, 'WorkingFiles/q2info.tsv'), index=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Generate Xq_minus_X.tsv\n",
    "#\n",
    "\n",
    "xq = [i.drop('idinf', axis=1) for i in xq]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of proteins in LD2214-TMT1: 5247\n",
      "Number of proteins in LD2214-TMT2: 4936\n",
      "Number of proteins in LD2223-TMT1: 5247\n",
      "Number of proteins in LD2223-TMT2: 4936\n"
     ]
    }
   ],
   "source": [
    "# Select between LD2214-TMT1 and LD2214-TMT2\n",
    "for i in xq:\n",
    "    if 'LD2214' in i.columns[0]:\n",
    "        print(f'Number of proteins in {i.columns[0]}: {i.shape[0]}')\n",
    "\n",
    "for i in xq:\n",
    "    if 'LD2223' in i.columns[0]:\n",
    "        print(f'Number of proteins in {i.columns[0]}: {i.shape[0]}')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select both from TMT1\n",
    "\n",
    "xq = [i for i in xq if i.columns[0] not in ['LD2214-TMT2', 'LD2223-TMT2']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "xq = reduce(lambda df1, df2: df1.join(df2, how='outer'), xq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "xq.columns = [i.split('-')[0] for i in xq.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "xq = xq.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "xq.to_csv(os.path.join(working_path, 'WorkingFiles/Xq_minus_X.tsv'), sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A0A023NE65</th>\n",
       "      <th>A0A068ETZ5</th>\n",
       "      <th>A0A075B5M7</th>\n",
       "      <th>A0A075B5N7</th>\n",
       "      <th>A0A075B5S2</th>\n",
       "      <th>A0A076FRG6</th>\n",
       "      <th>A0A077K846</th>\n",
       "      <th>A0A077S2U6</th>\n",
       "      <th>A0A087WNU5</th>\n",
       "      <th>A0A087WQ04</th>\n",
       "      <th>...</th>\n",
       "      <th>V9GX23</th>\n",
       "      <th>V9GX81</th>\n",
       "      <th>V9GZG5</th>\n",
       "      <th>V9GZG9</th>\n",
       "      <th>V9GZQ6</th>\n",
       "      <th>W8QK46</th>\n",
       "      <th>X5J4Q6</th>\n",
       "      <th>X5J4V5</th>\n",
       "      <th>X5J5L2</th>\n",
       "      <th>Z4YKV1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LD2169</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.211745</td>\n",
       "      <td>-0.279754</td>\n",
       "      <td>-0.148536</td>\n",
       "      <td>0.042719</td>\n",
       "      <td>0.374578</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.176766</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.083081</td>\n",
       "      <td>...</td>\n",
       "      <td>0.296407</td>\n",
       "      <td>0.412030</td>\n",
       "      <td>0.037165</td>\n",
       "      <td>0.078172</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.388321</td>\n",
       "      <td>0.635805</td>\n",
       "      <td>-0.124981</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2170</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.058881</td>\n",
       "      <td>-0.830888</td>\n",
       "      <td>-0.165629</td>\n",
       "      <td>-0.104145</td>\n",
       "      <td>0.173245</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.000786</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.073758</td>\n",
       "      <td>...</td>\n",
       "      <td>0.417525</td>\n",
       "      <td>0.481374</td>\n",
       "      <td>0.090895</td>\n",
       "      <td>0.069856</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.214174</td>\n",
       "      <td>0.085245</td>\n",
       "      <td>-0.551075</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2171</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.137228</td>\n",
       "      <td>-0.476768</td>\n",
       "      <td>0.773273</td>\n",
       "      <td>0.048258</td>\n",
       "      <td>-0.001902</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.116687</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.175519</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.299232</td>\n",
       "      <td>0.060244</td>\n",
       "      <td>0.080900</td>\n",
       "      <td>0.101585</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.072631</td>\n",
       "      <td>0.505704</td>\n",
       "      <td>-0.002342</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2243</th>\n",
       "      <td>0.001684</td>\n",
       "      <td>0.160198</td>\n",
       "      <td>-0.439884</td>\n",
       "      <td>0.010680</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.071086</td>\n",
       "      <td>0.175706</td>\n",
       "      <td>-0.214627</td>\n",
       "      <td>0.048655</td>\n",
       "      <td>...</td>\n",
       "      <td>0.154346</td>\n",
       "      <td>0.588669</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.079279</td>\n",
       "      <td>-0.048808</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.156649</td>\n",
       "      <td>-0.355584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2244</th>\n",
       "      <td>0.139485</td>\n",
       "      <td>-0.204659</td>\n",
       "      <td>-0.281001</td>\n",
       "      <td>0.193496</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.179846</td>\n",
       "      <td>-0.277834</td>\n",
       "      <td>0.029755</td>\n",
       "      <td>0.015017</td>\n",
       "      <td>...</td>\n",
       "      <td>0.220831</td>\n",
       "      <td>0.155407</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.078939</td>\n",
       "      <td>0.375996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.063778</td>\n",
       "      <td>-0.314014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2245</th>\n",
       "      <td>0.071298</td>\n",
       "      <td>-0.263406</td>\n",
       "      <td>-0.820109</td>\n",
       "      <td>-0.016369</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.020832</td>\n",
       "      <td>-0.010656</td>\n",
       "      <td>0.257595</td>\n",
       "      <td>0.075472</td>\n",
       "      <td>...</td>\n",
       "      <td>0.020431</td>\n",
       "      <td>-0.169871</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.071655</td>\n",
       "      <td>0.237942</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.094179</td>\n",
       "      <td>0.034142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2162</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.007486</td>\n",
       "      <td>-0.145293</td>\n",
       "      <td>0.179016</td>\n",
       "      <td>-0.121668</td>\n",
       "      <td>0.199740</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.321679</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.213393</td>\n",
       "      <td>...</td>\n",
       "      <td>0.246303</td>\n",
       "      <td>0.206115</td>\n",
       "      <td>-0.026002</td>\n",
       "      <td>-0.003327</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.563918</td>\n",
       "      <td>0.721845</td>\n",
       "      <td>-0.285067</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2163</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.260648</td>\n",
       "      <td>-0.607871</td>\n",
       "      <td>-0.069603</td>\n",
       "      <td>-0.184768</td>\n",
       "      <td>0.314014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.038032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.079964</td>\n",
       "      <td>...</td>\n",
       "      <td>0.384033</td>\n",
       "      <td>-0.076501</td>\n",
       "      <td>0.040769</td>\n",
       "      <td>-0.052532</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.369158</td>\n",
       "      <td>0.333702</td>\n",
       "      <td>-0.515202</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2164</th>\n",
       "      <td>-0.050325</td>\n",
       "      <td>-0.106601</td>\n",
       "      <td>-0.437660</td>\n",
       "      <td>0.070138</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.092557</td>\n",
       "      <td>0.343222</td>\n",
       "      <td>-0.417765</td>\n",
       "      <td>0.247348</td>\n",
       "      <td>...</td>\n",
       "      <td>0.138059</td>\n",
       "      <td>0.166713</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.049737</td>\n",
       "      <td>0.065082</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.290823</td>\n",
       "      <td>-0.258613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2214</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.031807</td>\n",
       "      <td>-0.297856</td>\n",
       "      <td>-0.145410</td>\n",
       "      <td>0.293554</td>\n",
       "      <td>0.015671</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.157770</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.062284</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.264564</td>\n",
       "      <td>0.380530</td>\n",
       "      <td>-0.209537</td>\n",
       "      <td>0.031008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.034763</td>\n",
       "      <td>0.389654</td>\n",
       "      <td>-0.328073</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2215</th>\n",
       "      <td>0.066992</td>\n",
       "      <td>-0.133346</td>\n",
       "      <td>0.106556</td>\n",
       "      <td>0.051030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.037885</td>\n",
       "      <td>-0.056545</td>\n",
       "      <td>0.217411</td>\n",
       "      <td>0.024338</td>\n",
       "      <td>...</td>\n",
       "      <td>0.139283</td>\n",
       "      <td>0.080192</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.098024</td>\n",
       "      <td>-0.085683</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.076450</td>\n",
       "      <td>0.035301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2166</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.084525</td>\n",
       "      <td>-0.256053</td>\n",
       "      <td>-0.055820</td>\n",
       "      <td>0.164855</td>\n",
       "      <td>0.281523</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.165856</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.000750</td>\n",
       "      <td>...</td>\n",
       "      <td>0.056975</td>\n",
       "      <td>0.222254</td>\n",
       "      <td>0.121860</td>\n",
       "      <td>-0.125362</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.433302</td>\n",
       "      <td>-0.223520</td>\n",
       "      <td>-0.720752</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2157</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.245650</td>\n",
       "      <td>-0.085077</td>\n",
       "      <td>0.119938</td>\n",
       "      <td>-0.052579</td>\n",
       "      <td>0.430747</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.190953</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.078767</td>\n",
       "      <td>...</td>\n",
       "      <td>0.035858</td>\n",
       "      <td>0.341728</td>\n",
       "      <td>0.060473</td>\n",
       "      <td>-0.057046</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.111636</td>\n",
       "      <td>0.422605</td>\n",
       "      <td>-0.145982</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2158</th>\n",
       "      <td>0.047752</td>\n",
       "      <td>-0.047914</td>\n",
       "      <td>0.216731</td>\n",
       "      <td>-0.132419</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.041100</td>\n",
       "      <td>0.014027</td>\n",
       "      <td>0.039245</td>\n",
       "      <td>-0.011406</td>\n",
       "      <td>...</td>\n",
       "      <td>0.109557</td>\n",
       "      <td>-0.041466</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.255236</td>\n",
       "      <td>-0.085730</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.164185</td>\n",
       "      <td>-0.171802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2223</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.043726</td>\n",
       "      <td>0.216452</td>\n",
       "      <td>-0.045655</td>\n",
       "      <td>0.122698</td>\n",
       "      <td>0.252574</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.001881</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.004507</td>\n",
       "      <td>...</td>\n",
       "      <td>0.258366</td>\n",
       "      <td>0.150145</td>\n",
       "      <td>-0.119268</td>\n",
       "      <td>-0.021207</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.445167</td>\n",
       "      <td>0.198969</td>\n",
       "      <td>0.035206</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2225</th>\n",
       "      <td>-0.119651</td>\n",
       "      <td>-0.301971</td>\n",
       "      <td>-0.064068</td>\n",
       "      <td>0.230250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.091213</td>\n",
       "      <td>0.187783</td>\n",
       "      <td>-0.065214</td>\n",
       "      <td>0.211757</td>\n",
       "      <td>...</td>\n",
       "      <td>0.094080</td>\n",
       "      <td>-0.011675</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.046739</td>\n",
       "      <td>0.019660</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.007888</td>\n",
       "      <td>0.005617</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16 rows × 6064 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        A0A023NE65  A0A068ETZ5  A0A075B5M7  A0A075B5N7  A0A075B5S2  \\\n",
       "LD2169         NaN    0.211745   -0.279754   -0.148536    0.042719   \n",
       "LD2170         NaN    0.058881   -0.830888   -0.165629   -0.104145   \n",
       "LD2171         NaN   -0.137228   -0.476768    0.773273    0.048258   \n",
       "LD2243    0.001684    0.160198   -0.439884    0.010680         NaN   \n",
       "LD2244    0.139485   -0.204659   -0.281001    0.193496         NaN   \n",
       "LD2245    0.071298   -0.263406   -0.820109   -0.016369         NaN   \n",
       "LD2162         NaN   -0.007486   -0.145293    0.179016   -0.121668   \n",
       "LD2163         NaN   -0.260648   -0.607871   -0.069603   -0.184768   \n",
       "LD2164   -0.050325   -0.106601   -0.437660    0.070138         NaN   \n",
       "LD2214         NaN   -0.031807   -0.297856   -0.145410    0.293554   \n",
       "LD2215    0.066992   -0.133346    0.106556    0.051030         NaN   \n",
       "LD2166         NaN    0.084525   -0.256053   -0.055820    0.164855   \n",
       "LD2157         NaN    0.245650   -0.085077    0.119938   -0.052579   \n",
       "LD2158    0.047752   -0.047914    0.216731   -0.132419         NaN   \n",
       "LD2223         NaN   -0.043726    0.216452   -0.045655    0.122698   \n",
       "LD2225   -0.119651   -0.301971   -0.064068    0.230250         NaN   \n",
       "\n",
       "        A0A076FRG6  A0A077K846  A0A077S2U6  A0A087WNU5  A0A087WQ04  ...  \\\n",
       "LD2169    0.374578         NaN    0.176766         NaN    0.083081  ...   \n",
       "LD2170    0.173245         NaN   -0.000786         NaN    0.073758  ...   \n",
       "LD2171   -0.001902         NaN   -0.116687         NaN   -0.175519  ...   \n",
       "LD2243         NaN    0.071086    0.175706   -0.214627    0.048655  ...   \n",
       "LD2244         NaN   -0.179846   -0.277834    0.029755    0.015017  ...   \n",
       "LD2245         NaN    0.020832   -0.010656    0.257595    0.075472  ...   \n",
       "LD2162    0.199740         NaN    0.321679         NaN    0.213393  ...   \n",
       "LD2163    0.314014         NaN    0.038032         NaN   -0.079964  ...   \n",
       "LD2164         NaN   -0.092557    0.343222   -0.417765    0.247348  ...   \n",
       "LD2214    0.015671         NaN    0.157770         NaN    0.062284  ...   \n",
       "LD2215         NaN   -0.037885   -0.056545    0.217411    0.024338  ...   \n",
       "LD2166    0.281523         NaN    0.165856         NaN   -0.000750  ...   \n",
       "LD2157    0.430747         NaN    0.190953         NaN    0.078767  ...   \n",
       "LD2158         NaN   -0.041100    0.014027    0.039245   -0.011406  ...   \n",
       "LD2223    0.252574         NaN   -0.001881         NaN   -0.004507  ...   \n",
       "LD2225         NaN    0.091213    0.187783   -0.065214    0.211757  ...   \n",
       "\n",
       "          V9GX23    V9GX81    V9GZG5    V9GZG9    V9GZQ6    W8QK46    X5J4Q6  \\\n",
       "LD2169  0.296407  0.412030  0.037165  0.078172       NaN -0.388321  0.635805   \n",
       "LD2170  0.417525  0.481374  0.090895  0.069856       NaN -0.214174  0.085245   \n",
       "LD2171 -0.299232  0.060244  0.080900  0.101585       NaN  0.072631  0.505704   \n",
       "LD2243  0.154346  0.588669       NaN -0.079279 -0.048808       NaN       NaN   \n",
       "LD2244  0.220831  0.155407       NaN -0.078939  0.375996       NaN       NaN   \n",
       "LD2245  0.020431 -0.169871       NaN -0.071655  0.237942       NaN       NaN   \n",
       "LD2162  0.246303  0.206115 -0.026002 -0.003327       NaN -0.563918  0.721845   \n",
       "LD2163  0.384033 -0.076501  0.040769 -0.052532       NaN -0.369158  0.333702   \n",
       "LD2164  0.138059  0.166713       NaN  0.049737  0.065082       NaN       NaN   \n",
       "LD2214 -0.264564  0.380530 -0.209537  0.031008       NaN  0.034763  0.389654   \n",
       "LD2215  0.139283  0.080192       NaN -0.098024 -0.085683       NaN       NaN   \n",
       "LD2166  0.056975  0.222254  0.121860 -0.125362       NaN -0.433302 -0.223520   \n",
       "LD2157  0.035858  0.341728  0.060473 -0.057046       NaN -0.111636  0.422605   \n",
       "LD2158  0.109557 -0.041466       NaN -0.255236 -0.085730       NaN       NaN   \n",
       "LD2223  0.258366  0.150145 -0.119268 -0.021207       NaN -0.445167  0.198969   \n",
       "LD2225  0.094080 -0.011675       NaN -0.046739  0.019660       NaN       NaN   \n",
       "\n",
       "          X5J4V5    X5J5L2    Z4YKV1  \n",
       "LD2169 -0.124981       NaN       NaN  \n",
       "LD2170 -0.551075       NaN       NaN  \n",
       "LD2171 -0.002342       NaN       NaN  \n",
       "LD2243       NaN -0.156649 -0.355584  \n",
       "LD2244       NaN -0.063778 -0.314014  \n",
       "LD2245       NaN -0.094179  0.034142  \n",
       "LD2162 -0.285067       NaN       NaN  \n",
       "LD2163 -0.515202       NaN       NaN  \n",
       "LD2164       NaN -0.290823 -0.258613  \n",
       "LD2214 -0.328073       NaN       NaN  \n",
       "LD2215       NaN  0.076450  0.035301  \n",
       "LD2166 -0.720752       NaN       NaN  \n",
       "LD2157 -0.145982       NaN       NaN  \n",
       "LD2158       NaN -0.164185 -0.171802  \n",
       "LD2223  0.035206       NaN       NaN  \n",
       "LD2225       NaN -0.007888  0.005617  \n",
       "\n",
       "[16 rows x 6064 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Start Standardization and Exploratory Data Analysis\n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "mdata = pd.read_csv(r\"S:\\U_Proteomica\\UNIDAD\\software\\MacrosRafa\\data\\Metabolomics\\PESA_Integromics\\Data\\Metadata\\ALDH4\\WorkingFiles\\main_metadata.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Check missing values\n",
    "#\n",
    "\n",
    "plotMV = PlotMV(xq, mdata, file=fileSummary)\n",
    "plotMV.plotSummary(col='Group', g1='A12', g2='B1-8', titleLabel='A12 vs B1-8')\n",
    "plotMV.plotSummary(col='Group', g1='A12', g2='PBS', titleLabel='A12 vs PBS')\n",
    "plotMV.plotSummary(col='Group', g1='B1-8', g2='PBS', titleLabel='B1-8 vs PBS')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of observations: 16\n",
      "Total number of proteins: 6064\n",
      "Total number of proteins with <20.0% of missing values(<3 of obs.): 4118\n",
      "\n",
      "Group  TMT \n",
      "A12    TMT1    3\n",
      "       TMT2    3\n",
      "B1-8   TMT1    3\n",
      "       TMT2    2\n",
      "PBS    TMT1    3\n",
      "       TMT2    2\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "MVF_thr = 0.2\n",
    "\n",
    "logw(f\"Total number of observations: {xq.shape[0]}\")\n",
    "logw(f\"Total number of proteins: {xq.shape[1]}\")\n",
    "logw(f\"Total number of proteins with <{MVF_thr*100}% of missing values(<{int(xq.shape[0]*MVF_thr)} of obs.): {((xq.isna().sum()/xq.shape[0])<=MVF_thr).sum()}\")\n",
    "\n",
    "logw(\"\")\n",
    "logw(mdata[np.isin(mdata['Seqn'], xq.index)].loc[:, ['Group', 'TMT']].groupby(['Group', 'TMT']).size())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Generate Xq_minus_X_norm.tsv\n",
    "# \n",
    "\n",
    "# Filter by 0.2 missing values\n",
    "xqf = xq.loc[:, (xq.isna().sum()/xq.shape[0])<MVF_thr].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total missing values: 5\n",
      "\n",
      "LD2170    3\n",
      "LD2163    1\n",
      "LD2214    1\n",
      "dtype: int64\n",
      "\n",
      "P97313    2\n",
      "P97822    1\n",
      "Q5EBK7    1\n",
      "Q8CI11    1\n",
      "dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>P97313</th>\n",
       "      <th>P97822</th>\n",
       "      <th>Q5EBK7</th>\n",
       "      <th>Q8CI11</th>\n",
       "      <th>Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LD2169</th>\n",
       "      <td>-0.271509</td>\n",
       "      <td>0.202039</td>\n",
       "      <td>0.281283</td>\n",
       "      <td>-0.163746</td>\n",
       "      <td>A12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2170</th>\n",
       "      <td>-0.172375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2171</th>\n",
       "      <td>0.005710</td>\n",
       "      <td>-0.675992</td>\n",
       "      <td>-0.233917</td>\n",
       "      <td>0.051111</td>\n",
       "      <td>A12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2243</th>\n",
       "      <td>0.049698</td>\n",
       "      <td>-0.185714</td>\n",
       "      <td>0.129795</td>\n",
       "      <td>-0.065201</td>\n",
       "      <td>A12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2244</th>\n",
       "      <td>0.322827</td>\n",
       "      <td>-0.109768</td>\n",
       "      <td>0.230855</td>\n",
       "      <td>0.793996</td>\n",
       "      <td>A12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2245</th>\n",
       "      <td>-0.400419</td>\n",
       "      <td>-0.291929</td>\n",
       "      <td>-0.013153</td>\n",
       "      <td>0.607114</td>\n",
       "      <td>A12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2162</th>\n",
       "      <td>-0.405263</td>\n",
       "      <td>0.200672</td>\n",
       "      <td>0.610763</td>\n",
       "      <td>-0.150073</td>\n",
       "      <td>B1-8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2163</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.306967</td>\n",
       "      <td>0.299875</td>\n",
       "      <td>0.215773</td>\n",
       "      <td>B1-8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2164</th>\n",
       "      <td>-0.542329</td>\n",
       "      <td>-0.219494</td>\n",
       "      <td>0.170170</td>\n",
       "      <td>-0.106408</td>\n",
       "      <td>B1-8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2214</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.253333</td>\n",
       "      <td>-0.046615</td>\n",
       "      <td>-0.353492</td>\n",
       "      <td>B1-8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2215</th>\n",
       "      <td>-0.112600</td>\n",
       "      <td>0.264657</td>\n",
       "      <td>0.088918</td>\n",
       "      <td>-0.176514</td>\n",
       "      <td>B1-8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2166</th>\n",
       "      <td>-0.356998</td>\n",
       "      <td>0.183562</td>\n",
       "      <td>0.030076</td>\n",
       "      <td>0.595803</td>\n",
       "      <td>PBS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2157</th>\n",
       "      <td>-0.094991</td>\n",
       "      <td>-0.205475</td>\n",
       "      <td>0.426152</td>\n",
       "      <td>0.017101</td>\n",
       "      <td>PBS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2158</th>\n",
       "      <td>-0.036213</td>\n",
       "      <td>-0.043354</td>\n",
       "      <td>-0.052586</td>\n",
       "      <td>0.562574</td>\n",
       "      <td>PBS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2223</th>\n",
       "      <td>0.135135</td>\n",
       "      <td>0.233021</td>\n",
       "      <td>0.109265</td>\n",
       "      <td>0.104990</td>\n",
       "      <td>PBS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LD2225</th>\n",
       "      <td>-0.383926</td>\n",
       "      <td>-0.114802</td>\n",
       "      <td>0.011602</td>\n",
       "      <td>0.538546</td>\n",
       "      <td>PBS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          P97313    P97822    Q5EBK7    Q8CI11 Group\n",
       "LD2169 -0.271509  0.202039  0.281283 -0.163746   A12\n",
       "LD2170 -0.172375       NaN       NaN       NaN   A12\n",
       "LD2171  0.005710 -0.675992 -0.233917  0.051111   A12\n",
       "LD2243  0.049698 -0.185714  0.129795 -0.065201   A12\n",
       "LD2244  0.322827 -0.109768  0.230855  0.793996   A12\n",
       "LD2245 -0.400419 -0.291929 -0.013153  0.607114   A12\n",
       "LD2162 -0.405263  0.200672  0.610763 -0.150073  B1-8\n",
       "LD2163       NaN  0.306967  0.299875  0.215773  B1-8\n",
       "LD2164 -0.542329 -0.219494  0.170170 -0.106408  B1-8\n",
       "LD2214       NaN  0.253333 -0.046615 -0.353492  B1-8\n",
       "LD2215 -0.112600  0.264657  0.088918 -0.176514  B1-8\n",
       "LD2166 -0.356998  0.183562  0.030076  0.595803   PBS\n",
       "LD2157 -0.094991 -0.205475  0.426152  0.017101   PBS\n",
       "LD2158 -0.036213 -0.043354 -0.052586  0.562574   PBS\n",
       "LD2223  0.135135  0.233021  0.109265  0.104990   PBS\n",
       "LD2225 -0.383926 -0.114802  0.011602  0.538546   PBS"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f'Total missing values: {xqf.isna().sum().sum()}')\n",
    "print()\n",
    "print(xqf.isna().sum(axis=1)[xqf.isna().any(axis=1)])\n",
    "print()\n",
    "print(xqf.isna().sum(axis=0)[xqf.isna().any(axis=0)])\n",
    "\n",
    "xqf.loc[:, xqf.isna().any(axis=0)].join(mdata.set_index('Seqn').loc[:, ['Group']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Center and scale\n",
    "#\n",
    "\n",
    "xqfn = pd.DataFrame(\n",
    "    StandardScaler().fit_transform(xqf),\n",
    "    columns=xqf.columns, index=xqf.index\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Summary plots\n",
    "#\n",
    "\n",
    "plotEDA = PlotEDA(xqf, mdata, file=fileSummary)\n",
    "\n",
    "plotEDA.plotSummary(r11=(-0.4,0.4), r12=(0,2), r21=(-0.4,0.4), r22=(0,2), r3=(-4,4),vl3=[0])\n",
    "plotEDA.plotByGroup('Group', vl1=[0],vl2=[0])\n",
    "plotEDA.plotByGroup('Ig', vl1=[0],vl2=[0])\n",
    "plotEDA.plotByGroup('Control', vl1=[0],vl2=[0])\n",
    "plotEDA.plotByGroup('TMT',vl1=[0],vl2=[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Imputation of missing values using KNN\n",
    "#\n",
    "\n",
    "from sklearn.impute import KNNImputer\n",
    "\n",
    "imputer = KNNImputer(n_neighbors=3)\n",
    "xqfnv = pd.DataFrame(\n",
    "    imputer.fit_transform(xqfn),\n",
    "    columns=xqfn.columns,\n",
    "    index=xqfn.index\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of values: 65888\n",
      "Number of missing values imputed by KNN (n=3): 5 (0.01%)\n"
     ]
    }
   ],
   "source": [
    "logw(f\"Total number of values: {xqfn.shape[0]*xqfn.shape[1]}\")\n",
    "logw(f\"Number of missing values imputed by KNN (n=3): {xqfn.isna().sum().sum()} ({round(xqfn.isna().sum().sum()/(xqfn.shape[0]*xqfn.shape[1])*100,2)}%)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading required package: mgcv\n",
      "Loading required package: nlme\n",
      "This is mgcv 1.8-41. For overview type 'help(\"mgcv-package\")'.\n",
      "Loading required package: genefilter\n",
      "Loading required package: BiocParallel\n",
      "Warning messages:\n",
      "1: package 'sva' was built under R version 4.2.1 \n",
      "2: package 'mgcv' was built under R version 4.2.2 \n",
      "3: package 'nlme' was built under R version 4.2.2 \n",
      "4: package 'genefilter' was built under R version 4.2.2 \n",
      "5: package 'BiocParallel' was built under R version 4.2.2 \n",
      "Found2batches\n",
      "Adjusting for2covariate(s) or covariate level(s)\n",
      "Standardizing Data across genes\n",
      "Fitting L/S model and finding priors\n",
      "Finding nonparametric adjustments\n",
      "Adjusting the Data\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from myComBat import myComBat\n",
    "\n",
    "catVars = ['Group']\n",
    "conVars = []\n",
    "xqfnvb = myComBat(xqfnv, mdata, 'TMT', catVars, conVars, Rpath=os.path.join(working_path, 'WorkingFiles', 'myRData'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotEDA = PlotEDA(xqfnvb, mdata, file=fileSummary)\n",
    "plotEDA.plotByGroup('Group', vl1=[0],vl2=[0], plotN=True, titleLabel='- Corrected by Batch Effect')\n",
    "plotEDA.plotByGroup('Ig', vl1=[0],vl2=[0], plotN=True, titleLabel='- Corrected by Batch Effect')\n",
    "plotEDA.plotByGroup('Control', vl1=[0],vl2=[0], plotN=True, titleLabel='- Corrected by Batch Effect')\n",
    "plotEDA.plotByGroup('TMT',vl1=[0],vl2=[0], plotN=True, titleLabel='- Corrected by Batch Effect')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "ploteda = PlotEDA(xqfnvb, mdata, file=fileSummary)\n",
    "ploteda.plotSummary(plots=[1], titleLabel='- Corrected by Batch Effect', r11=(-0.5, 0.5), r12=(0.85, 1.15))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Kuskal-Wallis test\n",
      "Non corrected\n",
      "Group - KruskalResult(statistic=10.0858385618194, pvalue=0.006454877186878417)\n",
      "TMT - KruskalResult(statistic=5.113835604017368, pvalue=0.02373578950885407)\n",
      "Corrected\n",
      "Group - KruskalResult(statistic=13.134777416737052, pvalue=0.001405462697469717)\n",
      "TMT - KruskalResult(statistic=4.190686121553881, pvalue=0.040646643522018834)\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Kruskal-Wallisk H test \n",
    "#\n",
    "\n",
    "logw('')\n",
    "logw('Kuskal-Wallis test')\n",
    "logw('Non corrected')\n",
    "logw(f'Group - {plotEDA._kruskal(xqfnv, \"Group\")}')\n",
    "logw(f'TMT - {plotEDA._kruskal(xqfnv, \"TMT\")}')\n",
    "logw('Corrected')\n",
    "logw(f'Group - {plotEDA._kruskal(xqfnvb, \"Group\")}')\n",
    "logw(f'TMT - {plotEDA._kruskal(xqfnvb, \"TMT\")}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "xqfnvb.to_csv(os.path.join(working_path, 'WorkingFiles', 'Xq_minus_X_norm.tsv'), sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "pcaumap = PCA_UMAP(xqfnv, mdata, file=filePCA)\n",
    "pcaumap.plotReduction('TMT', pcacomp=[0,1])\n",
    "pcaumap.plotReduction('Group', pcacomp=[0,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "pcaumap = PCA_UMAP(xqfnvb, mdata, file=filePCA)\n",
    "#pcaumap.plotReduction('TMT', pcacomp=[0,1], titleLabel='- Corrected by Batch Effect')\n",
    "pcaumap.plotReduction('Group', pcacomp=[2,4], titleLabel='- Corrected by Batch Effect')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>%Var PCA</th>\n",
       "      <th>Group</th>\n",
       "      <th>Ig</th>\n",
       "      <th>Control</th>\n",
       "      <th>TMT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22.877066</td>\n",
       "      <td>0.6796</td>\n",
       "      <td>0.9041</td>\n",
       "      <td>0.4908</td>\n",
       "      <td>0.1499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17.974089</td>\n",
       "      <td>0.4696</td>\n",
       "      <td>0.9012</td>\n",
       "      <td>0.3242</td>\n",
       "      <td>0.0442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11.553285</td>\n",
       "      <td>0.0996</td>\n",
       "      <td>0.0443</td>\n",
       "      <td>0.0911</td>\n",
       "      <td>0.8061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8.735733</td>\n",
       "      <td>0.3759</td>\n",
       "      <td>0.2810</td>\n",
       "      <td>0.8417</td>\n",
       "      <td>0.0018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7.582811</td>\n",
       "      <td>0.8900</td>\n",
       "      <td>0.9331</td>\n",
       "      <td>0.7124</td>\n",
       "      <td>0.6119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5.359510</td>\n",
       "      <td>0.1066</td>\n",
       "      <td>0.5026</td>\n",
       "      <td>0.0354</td>\n",
       "      <td>0.8794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5.056408</td>\n",
       "      <td>0.3630</td>\n",
       "      <td>0.1588</td>\n",
       "      <td>0.6632</td>\n",
       "      <td>0.3892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4.344536</td>\n",
       "      <td>0.3081</td>\n",
       "      <td>0.8573</td>\n",
       "      <td>0.2284</td>\n",
       "      <td>0.8113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3.576064</td>\n",
       "      <td>0.7094</td>\n",
       "      <td>0.5311</td>\n",
       "      <td>0.8815</td>\n",
       "      <td>0.9329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2.840511</td>\n",
       "      <td>0.0333</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.1975</td>\n",
       "      <td>0.9906</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     %Var PCA   Group      Ig  Control     TMT\n",
       "1   22.877066  0.6796  0.9041   0.4908  0.1499\n",
       "2   17.974089  0.4696  0.9012   0.3242  0.0442\n",
       "3   11.553285  0.0996  0.0443   0.0911  0.8061\n",
       "4    8.735733  0.3759  0.2810   0.8417  0.0018\n",
       "5    7.582811  0.8900  0.9331   0.7124  0.6119\n",
       "6    5.359510  0.1066  0.5026   0.0354  0.8794\n",
       "7    5.056408  0.3630  0.1588   0.6632  0.3892\n",
       "8    4.344536  0.3081  0.8573   0.2284  0.8113\n",
       "9    3.576064  0.7094  0.5311   0.8815  0.9329\n",
       "10   2.840511  0.0333  0.0078   0.1975  0.9906"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCA_Var(xqfnv, mdata, conVars, catVars+['Ig', 'Control','TMT'], n_comp=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>%Var PCA</th>\n",
       "      <th>Group</th>\n",
       "      <th>Ig</th>\n",
       "      <th>Control</th>\n",
       "      <th>TMT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23.738437</td>\n",
       "      <td>0.7862</td>\n",
       "      <td>0.8567</td>\n",
       "      <td>0.6275</td>\n",
       "      <td>0.7737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17.284722</td>\n",
       "      <td>0.1174</td>\n",
       "      <td>0.7148</td>\n",
       "      <td>0.1331</td>\n",
       "      <td>0.5402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13.428600</td>\n",
       "      <td>0.0315</td>\n",
       "      <td>0.0108</td>\n",
       "      <td>0.0641</td>\n",
       "      <td>0.9430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8.251975</td>\n",
       "      <td>0.9496</td>\n",
       "      <td>0.9605</td>\n",
       "      <td>0.7626</td>\n",
       "      <td>0.9464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6.117657</td>\n",
       "      <td>0.0789</td>\n",
       "      <td>0.6826</td>\n",
       "      <td>0.1073</td>\n",
       "      <td>0.8569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6.085620</td>\n",
       "      <td>0.1574</td>\n",
       "      <td>0.0698</td>\n",
       "      <td>0.1306</td>\n",
       "      <td>0.6003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4.717784</td>\n",
       "      <td>0.4800</td>\n",
       "      <td>0.7896</td>\n",
       "      <td>0.3834</td>\n",
       "      <td>0.9162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3.881346</td>\n",
       "      <td>0.7006</td>\n",
       "      <td>0.4720</td>\n",
       "      <td>0.9834</td>\n",
       "      <td>0.9181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3.369429</td>\n",
       "      <td>0.0971</td>\n",
       "      <td>0.0276</td>\n",
       "      <td>0.2865</td>\n",
       "      <td>0.9121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2.863820</td>\n",
       "      <td>0.7634</td>\n",
       "      <td>0.6566</td>\n",
       "      <td>0.4557</td>\n",
       "      <td>0.7888</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     %Var PCA   Group      Ig  Control     TMT\n",
       "1   23.738437  0.7862  0.8567   0.6275  0.7737\n",
       "2   17.284722  0.1174  0.7148   0.1331  0.5402\n",
       "3   13.428600  0.0315  0.0108   0.0641  0.9430\n",
       "4    8.251975  0.9496  0.9605   0.7626  0.9464\n",
       "5    6.117657  0.0789  0.6826   0.1073  0.8569\n",
       "6    6.085620  0.1574  0.0698   0.1306  0.6003\n",
       "7    4.717784  0.4800  0.7896   0.3834  0.9162\n",
       "8    3.881346  0.7006  0.4720   0.9834  0.9181\n",
       "9    3.369429  0.0971  0.0276   0.2865  0.9121\n",
       "10   2.863820  0.7634  0.6566   0.4557  0.7888"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCA_Var(xqfnvb, mdata, conVars, catVars+['Ig', 'Control','TMT'], n_comp=10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "299fabdee10379681b2207a83aa9f93c313ee5d5504e286e44b436bd7e45d8f9"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
